Beautify readme

FirefoxMetzger opened the pull request:
Closes: https://github.com/pygfx/pygfx/issues/423

This PR updates the readme and makes it more presentable. It also adds the logo, which is used as favicon in the docs.

Todo:
- [x] finish refactoring later sections of the readme
- [x] add logo
- [x] make logo favicon in the sphinx docs

FirefoxMetzger commented:
@Korijn Looking at the sections below the "under construction" note, I get the feeling that most of them should (and do) live somewhere in the docs already. I also don't think they are particularly relevant with the exception of the Introduction which highlights some features.

Are you okay with me liberally deleting the sections that I think won't add much value to new people finding the project and who have little familiarity with internals of rendering engines?

Korijn commented:
Yes, absolutely. You can also consider breaking out parts to another file especially for contributors (e.g. the section about testing examples).

Korijn commented on README.md:
````diff
@@ -1,11 +1,60 @@
+pip install -U pygfx glfw
+```
+
+If you already have a wgpu-compatible canvas (e.g., Qt, pyside, wx) installed
+you may choose to not install glfw. Similarly, if you only wish to use offscreen
+rendering (eg. in CI) you will also not need glfw.
````
Maybe also mention jupyter notebook setup?

almarklein replied:
Since glfw is really light there is not a problem if its installed "extra", so I'd put less emphasis on the not-installing part. Something along the lines of:

```suggestion
The glfw library provides a lightweight window to render to. It's not strictly needed
if you only use pygfx from the notebook, use offscreen rendering, or plan to use 
another wgpu-compatible canvas such as PyQt, PySide, or wx.
```

FirefoxMetzger commented:
@almarklein I am happy with the new readme for now. There is one discussion point on which features to highlight, but otherwise it should all be done.

RTR (and I will look into adding the logo in the mean time).

Korijn commented:
Looks like the sponza lighting still needs some more tweaking:

![image](https://user-images.githubusercontent.com/1882046/214838121-afba444d-3f0b-43ca-982a-8d1a756a8afe.png)

Korijn commented on README.md:
````diff
@@ -1,132 +1,135 @@
-We take a lot of inspiration from ThreeJS, e.g.:
+```bash
+pip install -U pygfx glfw
+```
````
all the same caveats apply as for wgpu-py, e.g. you need pip >=20.3 on linux, etc. I think that's going to lead to the same kind of issues on this repo if we don't mention it like we do over there

FirefoxMetzger replied:
I added a link to the platform requirements for wgpu-py and a general note.

Korijn commented on README.md:
```diff
@@ -1,132 +1,135 @@
+git clone <address_of_your_fork>
+cd pygfx
+git remote add upstream git@github.com:pygfx/pygfx.git
+pip install -e .[dev,docs,examples]
```
Are contributors smart enough to use a venv themselves?

FirefoxMetzger replied:
I agree there are many existing ones, but I don't know a good one (never searched). If you know one I'd be happy to link it. It's mainly for people that are new to open-source work to have something easy to get them started.

Alternatively, we could drop the contributing section entirely or shrink it. Depends on how eager we are to attract external contributions.

almarklein replied:
This one perhaps? https://docs.github.com/en/get-started/quickstart/github-flow

FirefoxMetzger replied:
Sounds good. I will still keep a snippet on how to get a dev install, because that is specific to pygfx (i.e. which optional dependencies to install).

Korijn commented:
I'd recommend to add very prominent links to the User Guide, API Reference and Example Gallery near the top

Korijn commented on README.md:
```diff
@@ -1,132 +1,135 @@
-See also the examples, they all do something like this:
+- Native support for High Resolution screens.
+- Builtin anti-aliasing.
+- Support for picking objects and parts within objects.
```
Let's make that "Builtin picking"

Korijn commented on README.md:
```diff
@@ -1,132 +1,135 @@
+- Custom post-processing. (do the examples for this need updating?)
+- Easy shader templating
+- built atop [WGPU](https://github.com/pygfx/wgpu-py) (>= OpenGL)
+- semantically similar to TreeJS (WorldObjects, decoupled rendering backend, ...)
```
I think the event system is also worth a mention

almarklein replied:
Idea: perhaps we can split out features particular of interest for 1) object rendering / meshes; 2) scientific viz.

FirefoxMetzger replied:
I would only list the 3 or 5 features that we are most proud of (or that sell the best) otherwise it quickly becomes a boring cloth-line.

For all the others, we should already have a feature example that shows off each feature right? So we can just point users to the gallery and say "check if the feature you want is listed here".

FirefoxMetzger replied:
I've made a selection of the top 5 features from all our suggestions. I kicked out some (light/shaddows, PBR, ...) that you could already see in the example images at the top, or that is mentioned elsewhere (TreeJS/WGPU) and then choose the 5 items that sounded the most impressive.

Suggestions are, ofc, welcome, and we also have to find the corresponding examples for each feature.

almarklein reviewed (commented):
Perhaps we should also include a section on the goals / purpose. Something like the following (feel free to rewrite in your own words):

Pygfx is a graphics library that leverages the GPU (via wgpu-py). It can be used for classical graphics (e.g. meshes, PBR rendering, lighting, shadows, etc.) as well as scientific visualizations.

From the point of view of scientific visualization, pygfx provides an API that can be considered relatively low-level. This is a deliberate choice to achieve flexibility; higher-level API's can be built on top of it, such as [fastplotlib](https://github.com/kushalkolar/fastplotlib).

Using the GPU brings high performance, but this does not mean speed is our top priority; we favor hackability and correctness. If you want a top tier render engine, you should probably not use Python anyway 😉

FirefoxMetzger commented:
@Korijn Can we get a version of the logo that combines the logo (on the left) and the name of the project (on the right)?

Korijn commented:
> @Korijn Can we get a version of the logo that combines the logo (on the left) and the name of the project (on the right)?

I can request it but it won't be ready anytime soon :)

FirefoxMetzger commented:
> I can request it but it won't be ready anytime soon :)

That's fine. As long as it is going through the pipeline we can replace it eventually ⭐ 

> I'll see if I can put something together myself later tonight. Any specifics like resolution I should take into account?

A landscape format would be good. If you want a specific solution, `600x300` is a good size. 

I also just realized that we can use svg in the readme, so I will kick out the png :)

Korijn commented:
Here you go:

![Banner](https://user-images.githubusercontent.com/1882046/214925024-36009d1e-35fc-4cb0-99de-2a07cac515c4.svg)

almarklein commented:
I tried to compose a "what is pygfx" that includes most of the features, which may be less tedious than a list of bullets. We may still cut some out for brevity though. Thoughts?

### What is pygfx

Pygfx is a graphics library that leverages the GPU via [WGPU](https://github.com/pygfx/wgpu-py) (the successor to OpenGL), and has an API heavily influenced by [ThreeJS](https://threejs.org/).

It can be used as a **render engine**, with support for physically based rendering, lights and shadows, order-independent transparenc, and more.

It also supports **scientific visualiations** with features like SDF text, colormaps, volume rendering, mesh- and volume slices, and the possibility to write custom shaders.

Interactive applications benefit from a powerful event system and picking mechanism.


### What it's not

From the point of view of scientific visualization, pygfx provides an API that can be considered relatively low-level. This is a deliberate choice to achieve flexibility; higher-level API's can be built on top of it, such as fastplotlib.

Using the GPU brings high performance, but this does not mean speed is our top priority; we favor hackability and correctness. If you want a top tier render engine, you should probably not use Python anyway 😉

FirefoxMetzger commented:
> Here you go (two font size choices):

Thanks @Korijn . I found the default Inkscape font a bit simple, so I spend some time looking for a serif font that could work. This is what I've settled for, let me know if you prefer this one or if we should go with the logo you've shared.

![pygfx_with_name](https://user-images.githubusercontent.com/4402489/215323523-8174fd01-28ee-4752-88b3-0df4dd9cf838.svg)

> I tried to compose a "what is pygfx" that includes most of the features, which may be less tedious than a list of bullets. We may still cut some out for brevity though. Thoughts?

I like it. I will put it into a "About pygfx" section which can live in the middle of the readme. I would keep this separate from the feature highlights though, because it is less of an "elevator pitch" and more of a "I'm hooked give me some details".

FirefoxMetzger commented:
Okay, this should resolve everything that was pending on this PR. A few things that need to be reviewed in this round:

- [ ] The logo that I added to the top of the readme
- [ ] The 5 feature highlights + links to examples
- [ ] the about section that I've added (+ maybe a citation for the project if we have/care)

Otherwise this PR should be ready.

almarklein commented on README.md:
```diff
@@ -1,132 +1,127 @@
-## Introduction
+To work correctly, pygfx needs _some_ window to render to. Glfw is one
+lightweight option, but there are others, too. If you only use pygfx for
+offscreen, or notebook rendering you may choose to omit glfw. Similarly, you can
```
```suggestion
offscreen or notebook rendering you may choose to omit glfw. Similarly, you can
```

almarklein commented on README.md:
```diff
@@ -1,132 +1,127 @@
+## Feature Highlights
+Some of pygfx's key features are:
 
+- SDF based rendering (example)
```
```suggestion
- SDF based text rendering (example)
```

almarklein commented on README.md:
```diff
@@ -1,132 +1,127 @@
 
+- SDF based rendering (example)
+- order-independent transparency (OIT) (example)
+- a shader templating system (example)
```
The shader templating is cool, but only interesting for ppl that want to extend pygfx. Let's replace with lights etc.
```suggestion
- lights, shadows, and physically based rendering (PBR)
```

almarklein commented on README.md:
```diff
@@ -1,132 +1,127 @@
+## Contributing
+We use a pull request (PR) based workflow similar to many other open-source
+libraries in the python ecosystem. You can read more about this workflow
+[here](https://docs.github.com/en/get-started/quickstart/github-flow); however,
```
The next sentence is more a consequence than a contradiction
```suggestion
[here](https://docs.github.com/en/get-started/quickstart/github-flow);
```

almarklein commented on README.md:
```diff
@@ -1,132 +1,127 @@
+## About pygfx
 
+Pygfx is a graphics library that uses WGPU (the successor of OpenGL) to provide
+GPU acceleration to rendering workloads. It mimics the ThreeJS API and is geared
```
It's more "inspired by" than "mimics", because we (deliberately) in some cases :)

almarklein commented on README.md:
```diff
@@ -1,132 +1,127 @@
 
+Pygfx is a graphics library that uses WGPU (the successor of OpenGL) to provide
+GPU acceleration to rendering workloads. It mimics the ThreeJS API and is geared
+towards scientific and medical visualization. At the same time, is mature enough
+to serve as a general-purpose rendering engine. (Yes, you _can_ write a game
```
Can we flip this around - first mention it's general purpose, then say it's geared towards scientific vis?

FirefoxMetzger replied:
swapped

almarklein commented on README.md:
```diff
@@ -1,132 +1,127 @@
+a GPU accelerated library.
 
-## On world objects, materials, and geometry
+@almarklein: ? citation ?
```
What kind of citation do you mean?

almarklein replied:
Let's skip for now.

almarklein reviewed (commented):
I made a few suggestions.

FirefoxMetzger commented on README.md:
```diff
@@ -1,132 +1,127 @@
-* Define an  animate function that calls: `renderer.render(scene, camera)`
+## About pygfx
 
+Pygfx is a graphics library that uses WGPU (the successor of OpenGL) to provide
```
@ivoflipse This one is at the beginning of a new sentence (paragraph in fact) so I capitalized it. It feels strange not to capitalize, though we can ofc use lowercase if you prefer.

ivoflipse replied:
What does scikit-learn do for example?

ivoflipse replied:
It seems to vary, which probably depends on who wrote it

https://scikit-learn.org/stable/faq.html

FirefoxMetzger replied:
On second thought, let's avoid starting a long discussion over something minor. I'll change it.

FirefoxMetzger commented on README.md:
```diff
@@ -1,132 +1,127 @@
+GPU acceleration to rendering workloads. It mimics the ThreeJS API and is geared
+towards scientific and medical visualization. At the same time, is mature enough
+to serve as a general-purpose rendering engine. (Yes, you _can_ write a game
+with it.) Thanks to it's low level of abstraction it is flexible and can be
```
fixed

Korijn commented:
Last one from me: `jupyter_rfb` is required for notebook usage, let's include that under "Installation"

almarklein commented on README.md:
```diff
@@ -1,132 +1,132 @@
+  https://pygfx.readthedocs.io/en/latest/_gallery/feature_demo/pbr.html))
+- event system with built-in picking ([example](
+  https://pygfx.readthedocs.io/en/latest/_gallery/feature_demo/picking_points.html))
+- Texture and color mapping supporting 1D, 2D and **3D data** ([example](
```
Consistency.
```suggestion
- texture and color mapping supporting 1D, 2D and 3D data ([example](
```

FirefoxMetzger replied:
nice catch.

FirefoxMetzger commented:
> Looks like the sponza lighting still needs some more tweaking:

Do we want to work out what tweaks sponza needs before we merge? 


If not, I think we are LGTM unless one of you knows a better example for a feature highlight compared to the one I chose.

Korijn commented:
Nah, I'm good with merging at this point. Let's create follow up issues for any left overs.

FirefoxMetzger commented:
@Korijn Shouldn't you be on vacation? 😅 We can let the issue sit here and you can merge it and we can deal with any consequences of that once you return.

almarklein commented on README.md:
```diff
@@ -1,132 +1,130 @@
+PyQt, PySide, or wx.
 
-We take a lot of inspiration from ThreeJS, e.g.:
+In addition there are some [platform
+requirements](https://github.com/pygfx/wgpu-py#platform-requirements). In
+essense, you need modern (enough) graphics drivers, and `pip>=20.3`.
```
These platform requirements in wgpu-py have moved into its docs.

```suggestion
In addition there are some platform
requirements, see the [wgpu docs](https://wgpu-py.readthedocs.io/en/stable/start.html). In
essence, you need modern (enough) graphics drivers, and `pip>=20.3`.
```

almarklein commented:
This is ready too @Korijn  :)

FirefoxMetzger commented:
RTR :)

Korijn commented:
Can you create an issue to track any remaining discussion items?

FirefoxMetzger commented:
@Korijn I think only the sponza scene situation is left. Otherwise I think we resolved everything that came up here.
